plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java

```diff
@@ -391,96 +449,112 @@ public void eventOccurred(RequestCompletedEvent eventArg) {
         }
     }
 
-    /**
-     * Building our own input stream instead of using the SDK's {@link com.microsoft.azure.storage.blob.BlobInputStream}
-     * because that stream is highly inefficient in both memory and CPU use.
-     */
-    private static class BlobInputStream extends InputStream {
+    private static class AzureInputStream extends StorageInputStream {
+        private final BlobAsyncClient client;
+        private final ByteBuffer buffer;
+        private final int maxRetries;
+        private final long firstReadOffset;
+        private final int firstReadLength;
+
+        private AzureInputStream(final BlobAsyncClient client,
+                                 long rangeOffset,
+                                 Long rangeLength,
+                                 int chunkSize,
+                                 long contentLength,
+                                 int maxRetries) throws IOException {
+            super(rangeOffset, rangeLength, chunkSize, contentLength);
+            this.client = client;
+            this.buffer = ByteBuffer.allocate(chunkSize);
```

Comment:
We can size this the same way we size `firstReadLength` so that we don't allocate 4M for a file smaller than 4M right?